circular-buffer/src/lib.rs

```diff
@@ -0,0 +1,818 @@
+
+    /// A slice that owns its elements, but not their storage. This is useful for things like
+    /// `Vec::retain` and `CircularBuffer::pop_slice`, since these functions can return a slice but
+    /// the elements of these slices would be leaked after the slice goes out of scope. `OwnedSlice`
+    /// simply manually drops all its elements when it goes out of scope.
+    #[derive(Debug, Eq, PartialEq)]
+    pub struct OwnedSlice<'a, T: 'a>(&'a mut [T]);
+
+    /// Owning iterator for `OwnedSlice`.
+    pub struct IntoIter<'a, T: 'a> {
+        slice: ManuallyDrop<OwnedSlice<'a, T>>,
+        index: usize,
+    }
+
+    impl<'a, T> Iterator for IntoIter<'a, T> {
+        type Item = T;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            use std::ptr;
+
+            let index = self.index;
+
+            if index >= self.slice.len() {
+                return None;
+            }
+
+            self.index += 1;
+
+            unsafe { Some(ptr::read(&self.slice[index])) }
+        }
+    }
+
+    impl<'a, T: 'a> IntoIterator for OwnedSlice<'a, T> {
+        type Item = T;
+        type IntoIter = IntoIter<'a, T>;
+
+        fn into_iter(self) -> Self::IntoIter {
+            IntoIter {
+                slice: ManuallyDrop::new(self),
+                index: 0,
+            }
+        }
+    }
+
+    impl<'a, T: 'a> OwnedSlice<'a, T> {
+        /// Construct an owned slice from a mutable slice pointer.
+        ///
+        /// # Unsafety
+        /// You must ensure that the memory pointed to by `inner` will not be accessible after the
+        /// lifetime of the `OwnedSlice`.
+        pub unsafe fn new(inner: &'a mut [T]) -> Self {
+            OwnedSlice(inner)
+        }
+    }
+
+    impl<'a, T> AsRef<[T]> for OwnedSlice<'a, T> {
+        fn as_ref(&self) -> &[T] {
+            self.0
+        }
+    }
+
+    impl<'a, T> AsMut<[T]> for OwnedSlice<'a, T> {
+        fn as_mut(&mut self) -> &mut [T] {
+            self.0
+        }
+    }
+
+    impl<'a, T> Deref for OwnedSlice<'a, T> {
+        type Target = [T];
+
+        fn deref(&self) -> &Self::Target {
+            self.0
+        }
+    }
+
+    impl<'a, T> DerefMut for OwnedSlice<'a, T> {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            self.0
+        }
+    }
+
+    impl<'a, T> Drop for OwnedSlice<'a, T> {
+        fn drop(&mut self) {
+            use std::ptr;
+
+            for element in self.iter_mut() {
+                unsafe {
+                    ptr::drop_in_place(element);
+                }
+            }
+        }
+    }
+}
+
+/// A fixed-size FIFO queue with safe access to the backing storage.
+///
+/// This type allows access to slices of the backing storage, for efficient, safe circular buffers
+/// of bytes or other `Copy` types.
+#[derive(Debug)]
+pub struct CircularBuffer<B: Array> {
+    // This must be manually dropped, as some or all of the elements may be uninitialized
+    buffer: ManuallyDrop<B>,
```

Comment:
i am worried about the prospect of manually dropping uninitialized memory. it should rather drop all known-to-be-live elements in the destructor

Reply:
I mean that the value of `B` must be manually dropped, as the default drop will drop all elements. What I do in the destructor is destroy only the known-to-be-live elements, just as you say.